src/User.cpp

```diff
@@ -0,0 +1,120 @@
+#include "../include/Session.h"
+
+// User's related methods implementation.
+User::User(const std::string& name): history(), name(name) {}
+
+const std::string User::getName() const {
+    return name;
+}
+
+const std::vector<Watchable*> User::get_history() const {
+    return history;
+}
+
+void User::addToHistory(Watchable& watchable) {
+    history.push_back(&watchable);
+}
+
+bool User::isInHistory(const Watchable& watchable) {
+    return std::find(history.begin(), history.end(), &watchable) != history.end();
+}
+
+
+// Length Recommender User's methods implementation.
+LengthRecommenderUser::LengthRecommenderUser(const std::string& name): User(name), _avgWatchableLen(0.0f) {}
+
+void LengthRecommenderUser::updateAverage() {
+    // we assume, that when updateAverage is called, new watchable has already been added to the history
+    // vector. hence, we decrease by 1 history.size(), in the multiplication, and using history.size() in
+    // the division
+    _avgWatchableLen = (_avgWatchableLen * (history.size() - 1) + history.back()->getLength())
+            / history.size();
+}
+
+void LengthRecommenderUser::addToHistory(Watchable& watchable) {
+    history.push_back(&watchable);
+    //  update average according to newly watched
+    updateAverage();
+}
+
+Watchable* LengthRecommenderUser::getRecommendation(Session& s) {
+    std::vector<Watchable*> content = s.getContent();
+    // set min difference to be max value before iterating, and recommended watchable to be null
+    float minDiff = std::numeric_limits<float>::max();
+    Watchable* recommendedWatchable = nullptr;
+    //iterate over all content vector, and compare it to min diff found
+    float curDiff;
+    for (auto iter = content.begin(); iter != content.end(); iter++)
+    {
+        curDiff =  abs((*iter)->getLength() - _avgWatchableLen);
+        if (curDiff < minDiff && !isInHistory(**iter)) {
+            recommendedWatchable = *iter;
+            minDiff = curDiff;
+        }
+    }
+    return recommendedWatchable;
+}
+
+// Rerun Recommender User's methods implementation.
+RerunRecommenderUser::RerunRecommenderUser(const std::string& name):
+    User(name), _lastRecIdx(-1) {}
+
+Watchable* RerunRecommenderUser::getRecommendation(Session& s) {
+    if (_lastRecIdx == -1) {
+        return *(history.begin());
+    }
+    _lastRecIdx = (_lastRecIdx+1) % history.size();
+    return history[_lastRecIdx];
+}
+
+// Genre Recommender User's methods implementation.
+GenreRecommenderUser::GenreRecommenderUser(const std::string& name): User(name), _popularTagsMap(),
+                                                                        _popularTagsVector() {}
+
+void GenreRecommenderUser::updatePopularTags() {
+    std::vector<std::string> lastWatchedTags = history.back()->getTags();
```

Comment:
This one should be defined const too I think

Reply:
it returns a vector of strings and so I didn't want to limit the user. maybe you are right and then it should return a reference and not an object as well. it is probably more 'memory efficient'

Reply:
I don't think it matter that much.. getTags will enforce it